README.md
# templimiter

Limits system temperature by reading thermal files, throttling the cpu, and sending SIGSTOP and SIGCONT signals.

## Description

This software was written for systems with inadequate hardware cooling. By constantly monitoring system thermal files, it is able to respond to dangerous temperature levels by either throttling the CPU or sending SIGSTOP signals to the highest CPU-consuming processes. If the temperature is below a given threshold, templimiter will dethrottle the CPU to maximum speeds and send SIGCONT signals to the processes stopped by the program.

All configuration is established via the templimiter.conf file. Path is defined during the pre-make `./configure` step. Use the `--which-conf` option to find its location on your system. Please refer to the [configuration](#configuration) section for all settings.

This program is meant to be run as a persistent daemon and is already configured to use systemd if possible. It is advisable to run the program with `--debug` first in order to verify that no configuration issues are preventing execution __before__ enabling the daemon. The daemon is configured to relaunch on exit.

## Installation

### Prerequisites

+ Linux
+ __c++17__
+ pkg-config
  + Necessary for automatic systemd configuration.
  + Manual configuration of systemd directory can still be acheived without pkg-config via `./configure --with-systemdsystemunitdir=DIR`

### Getting the Source

Clone the repo (or download the source archive).

```bash
git clone https://github.com/jpcx/templimiter.git
cd templimiter
```

### Building

_Note: This software uses autotools for building. Due to an issue regarding timestamp preservation, `./configure` may not work immediately._

If make fails with a 'not found' message, either:

+ Use the source archive instead of cloning the repository (timestamps are preserved).
+ Or execute `sh scripts/fix-timestamps.sh` before running configure and make.

```bash
./configure
make
sudo make install
```

### Setup

```bash
# Edit the configuration, if desired
sudo nano $(templimiter --which-conf)
```

### Verification

_NOTE: Setting low and tight thresholds for the following debugging step is recommended._

```bash
# Check for errors
sudo templimiter --debug
# In another terminal, stimulate the CPU
dd if=/dev/zero of=/dev/null bs=8M status=progress
```

_NOTE: During debug mode, if the thresholds are set appropriately you should see output signifying that actions were taken to limit temperature. Use `^C` to exit._

### Daemon Initialization

```bash
# If it runs as expected, enable the service
sudo systemctl enable templimiter
```

## Usage

templimiter is meant to be run as a daemon; simply start the service if you are using systemd.

___Note: Proper debugging is advised before enabling; the daemon is set to relaunch on failure.___

```bash
sudo systemctl start templimiter.service
```

Functionality remains the same if running without systemd.

## Modes

templimiter has two major modes of operation: throttle and SIGSTOP.

### Throttle

Throttle mode modifies scaling_max_freq files in response to temperature changes, which sets a ceiling to CPU speed. There are two methods of doing this- scaling mode and minmax mode:

+ __Scaling mode:__
  + Reads the scaling_available_frequencies file for available frequencies that the CPU can scale to.
    + Each iteration above the throttle threshold temperature will decrease the CPU speeds by one step.
    + Each iteration below the dethrottle threshold temperature will increase the CPU speeds by one step.
  + _This is the most fluid method of temperature limiting, and is recommended if possible._
    + As such, it is enabled by default.
    + If the system does not have scaling_available_frequencies files, the mode will be disabled and warn the console.
+ __Minmax mode:__
  + Reads the cpuinfo_max_freq files and cpuinfo_min_freq files for information on maximum and minimum cpu frequencies.
  + If the temperature exceeds the throttle threshold temperature CPU speed will be set to the minimums.
  + If the temperature exceeds the dethrottle threshold temperature CPU speed will be set to the maximums.

### SIGSTOP

SIGSTOP mode sends SIGSTOP signals to non-whitelisted processes in response to temperatures. There are two configuration options related to the operation: stepwise SIGSTOP and stepwise SIGCONT.

__WARNING: This mode can cause a lot of problems in the system if the whitelist is not configured perfectly. It is recommended that the SIGSTOP threshold be set to a high value or not used. It is disabled by default.__

+ __stepwise_SIGSTOP = true:__
  + Loads all available processes and picks the highest non-whitelisted process for sending the signal. Repeats on each iteration above the SIGSTOP threshold.
+ __stepwise_SIGSTOP = false:__
  + Upon reaching the SIGSTOP threshold, stops all non-whitelisted processes at once.
+ __stepwise_SIGCONT = true:__
  + Upon decreasing past the SIGCONT threshold, sends a SIGCONT signal to the lowest cpu-consuming process in the list of processes that have been sent a SIGSTOP signal by templimiter.
+ __stepwise_SIGCONT = false:__
  + Upon decreasing past the SIGCONT threshold, sends a SIGCONT signal to _all_ of the processes sent a SIGSTOP signal by templimiter.

## Documentation

### Configuration

A default configuration file is loaded into the correct location during the installation steps. To use a custom configuration directory, use `--with-templimiterconfdir=DIR` when running `./configure`.

#### Syntax

Configuration values must not be broken by newlines, but the number of spaces between tags/values does not matter.

__Default configuration:__

```conf
log_file_path            /var/log/templimiter.log
whitelist_comm           dnsmasq systemd (sd-pam) startx xinit Xorg dbus-daemon rtkit-daemon at-spi-bus-laun at-spi2-registr wpa_supplicant dhcpcd systemd-journal lvmetad systemd-udevd upowerd systemd-timesyn systemd-machine firewalld systemd-logind polkitd haveged systemd-resolve systemd-network
whitelist_state          S D Z T t W X x K W P
whitelist_pgrp           0 1
whitelist_max_nice       -21
use_throttle             true
use_SIGSTOP              false
use_scaling_available    true
use_stepwise_SIGSTOP     true
use_stepwise_SIGCONT     false
matcher_thermal          /sys/devices/virtual/thermal/thermal_zone*/temp
matcher_scaling_max_freq /sys/devices/system/cpu/cpu*/cpufreq/scaling_max_freq
matcher_cpuinfo_max_freq /sys/devices/system/cpu/cpu*/cpufreq/cpuinfo_max_freq
matcher_cpuinfo_min_freq /sys/devices/system/cpu/cpu*/cpufreq/cpuinfo_min_freq
matcher_scaling_available_frequencies /sys/devices/system/cpu/cpu*/cpufreq/scaling_available_frequencies
temp_SIGSTOP             70000
temp_SIGCONT             66000
temp_throttle            66000
temp_dethrottle          60000
min_sleep                500
```

___Note: The execution pid is automatically added to the whitelist; the program should not stop itself.___

#### Available Tags

| Tag | Type | Description |
| --- | --- | --- |
| log_file_path | string | Location of the log file to append to or create if blank |
| whitelist_pid | int[] | List of pids to whitelist |
| whitelist_comm | string[] | List of comm values to whitelist (may use * matching) |
| whitelist_state | char[] | List of state values to whitelist |
| whitelist_ppid |  int[] | List of ppid values to whitelist |
| whitelist_pgrp | int[] | List of pgrp values to whitelist |
| whitelist_session |  int[] | List of session values to whitelist |
| whitelist_tty_nr |  int[] | List of tty_nr values to whitelist |
| whitelist_tpgid |  int[] | List of tpgid values to whitelist |
| whitelist_flags |  unsigned[] | List of flags values to whitelist |
| whitelist_max_nice | long | nice values at or below this value will be whitelisted |
| use_throttle | (true \|\| false) | Toggle for throttle mode |
| use_SIGSTOP | (true \|\| false) | Toggle for SIGSTOP mode |
| use_scaling_available | (true \|\| false) | Toggle for scaling mode |
| use_stepwise_SIGSTOP | (true \|\| false) | Toggle for stepwise SIGSTOP mode |
| use_stepwise_SIGCONT | (true \|\| false) | Toggle for stepwise SIGCONT mode |
| matcher_thermal | string | Thermal file matcher (use * matching) |
| matcher_scaling_max_freq | string | scaling_max_freq file matcher (use * matching) |
| matcher_cpuinfo_max_freq | string | cpuinfo_max_freq file matcher (use * matching) |
| matcher_cpuinfo_min_freq | string | cpuinfo_min_freq file matcher (use * matching) |
| matcher_scaling_available_frequencies | string | scaling_available_frequencies file matcher (use * matching) |
| temp_SIGSTOP | unsigned long | maximum temperature found at any sensor to trigger SIGSTOP |
| temp_SIGCONT | unsigned long | minimum temperature found at the HOTTEST sensor to trigger SIGCONT |
| temp_throttle | unsigned long | maximum temperature found at any sensor to trigger throttling |
| temp_dethrottle | unsigned long | minimum temperature found at the HOTTEST sensor to trigger dethrottling |
| min_sleep | unsigned int | minimum time (in milliseconds) between re-scan operations |

### Source Code

All code is documented using Doxygen. If you would like to read the source code docs please run `doxygen Doxyfile` in the package root directory.

## License

Copyright © Justin Collier  License GPLv3+: GNU GPL version 3 or later
<<https://gnu.org/licenses/gpl.html>>.

This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


ChangeLog
2019-02-15  Justin Collier  <jpcxist@gmail.com>

	* templimiter: Version 0.1.0 released.

2019-02-15  Justin Collier  <jpcxist@gmail.com>

	* src/templimiter/daemon/monitor.cc
	(is_below_max_speed_, is_above_min_speed_): 
	Changed min/max criteria to use the midpoint between two steps in order
	minimize false positives due to fluctuations.
	(dethrottle_next_higher_, throttle_next_lower_)
	(dethrottle_highest_, throttle_lowest_)
	(exec_dethrottle_, exec_throttle_, run, Monitor):
	Added expected frequency checking based on previously set values.

	* src/templimiter/daemon/monitor.h
	(Monitor): Created variables that track unexpected throttle behavior in order 
	to minimize overthrottling when the scaling_max_freq files are being modified 
	externally.
	(dethrottle_highest_, throttle_lowest_): New arg cur_speeds.
	Provided cur_speeds externally in response to new expected frequency checking
	(some cores may be linked in the system).

	* src/templimiter/tools/vector.h
	(max_n_elements): Created procedure.
	(min_n_elements): Created procedure.

	* README.md: Clarified prerequisites and installation instructions.

2019-02-16  Justin Collier  <jpcxist@gmail.com>

	* src/main.cc: Updated version documentation.
	* README.md: Fixed fix-timestamps.sh clarification issue.


src/templimiter/daemon/config.h
/*
    Copyright (c) 2019 Justin Collier
    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

/**
 * @file config.h
 * @author Justin Collier (jpcxist@gmail.com)
 * @brief Provides templimiter::daemon::Config class
 * @date created 2019-02-08
 * @date modified 2019-02-09
 */

#pragma once

#include <memory>
#include <string>
#include <vector>

#include "templimiter/error/config-error.h"
#include "templimiter/error/error.h"
#include "templimiter/io/file-collection.h"
#include "templimiter/io/file.h"
#include "templimiter/io/operations.h"
#include "templimiter/tools/string.h"
#include "templimiter/tools/vector.h"

namespace templimiter {

namespace daemon {

/** @brief Contains configuration settings and derived objects */
class Config {
 private:
  // Internal variables
  /** @brief String vector used to store loaded configuration text */
  std::vector<std::string> config_lines_;
  /** @brief Holds the pid of the running process */
  pid_t own_pid_;
  /** @brief Location of /proc/self/stat used to find own_pid_ */
  const std::string PROC_SELF_STAT_ = "/proc/self/stat";
  /** @brief Location of /proc/stat used to find cpu information */
  const std::string PROC_STAT_ = "/proc/stat";

  // Internal pre-derivation config values
  /** @brief Matcher to find thermal files */
  std::string matcher_thermal_ =
      "/sys/devices/virtual/thermal/thermal_zone*/temp";
  /** @brief Matcher to find scaling_max_freq files */
  std::string matcher_scaling_max_freq_ =
      "/sys/devices/system/cpu/cpu*/cpufreq/scaling_max_freq";
  /** @brief Matcher to find cpuinfo_max_freq files */
  std::string matcher_cpuinfo_max_freq_ =
      "/sys/devices/system/cpu/cpu*/cpufreq/cpuinfo_max_freq";
  /** @brief Matcher to find cpuinfo_min_freq files */
  std::string matcher_cpuinfo_min_freq_ =
      "/sys/devices/system/cpu/cpu*/cpufreq/cpuinfo_min_freq";
  /** @brief Matcher to find scaling_available_frequencies files */
  std::string matcher_scaling_available_frequencies_ =
      "/sys/devices/system/cpu/cpu*/cpufreq/scaling_available_frequencies";

  // Private component of config values
  /** @brief File to send logging information */
  std::string log_file_path_ = "/var/log/templimiter.log";
  /** @brief Prohibits sending SIGSTOP to processes of this pid */
  std::vector<pid_t> whitelist_pid_;
  /** @brief Prohibits sending SIGSTOP to processes of this comm */
  std::vector<std::string> whitelist_comm_ = {
      "dnsmasq",         "systemd",         "(sd-pam)",
      "startx",          "xinit",           "Xorg",
      "dbus-daemon",     "rtkit-daemon",    "at-spi-bus-laun",
      "at-spi2-registr", "wpa_supplicant",  "dhcpcd",
      "systemd-journal", "lvmetad",         "systemd-udevd",
      "upowerd",         "systemd-timesyn", "systemd-machine",
      "firewalld",       "systemd-logind",  "polkitd",
      "haveged",         "systemd-resolve", "systemd-network"};
  /** @brief Prohibits sending SIGSTOP to processes of this state */
  std::vector<char> whitelist_state_;
  /** @brief Prohibits sending SIGSTOP to processes of this ppid */
  std::vector<pid_t> whitelist_ppid_;
  /** @brief Prohibits sending SIGSTOP to processes of this pgrp */
  std::vector<int> whitelist_pgrp_;
  /** @brief Prohibits sending SIGSTOP to processes of this session */
  std::vector<int> whitelist_session_;
  /** @brief Prohibits sending SIGSTOP to processes of this tty_nr */
  std::vector<int> whitelist_tty_nr_;
  /** @brief Prohibits sending SIGSTOP to processes of this tpgid */
  std::vector<int> whitelist_tpgid_;
  /** @brief Prohibits sending SIGSTOP to processes of this flag */
  std::vector<uint> whitelist_flags_;
  /** @brief Prohibits sending SIGSTOP to processes of a nice lower than this */
  long whitelist_max_nice_ = -21;
  /** @brief Whether or not throttling should be enabled */
  bool use_throttle_ = true;
  /** @brief Whether or not sending SIGSTOP signals should be enabled */
  bool use_SIGSTOP_ = false;
  /**
   * @brief Whether or not scaling_available_frequencies or cpu_min_freq should
   * be used
   */
  bool use_scaling_available_ = false;
  /** @brief Whether or not to stop one process at a time */
  bool use_stepwise_SIGSTOP_ = true;
  /** @brief Whether or not to continue one process at a time */
  bool use_stepwise_SIGCONT_ = false;
  /** @brief Temperature to start sending SIGSTOP signals */
  u_long temp_SIGSTOP_ = 70000;
  /** @brief Temperature to start sending SIGCONT signals */
  u_long temp_SIGCONT_ = 66000;
  /** @brief Temperature to start CPU throttling */
  u_long temp_throttle_ = 66000;
  /** @brief Temperature to start CPU dethrottling */
  u_long temp_dethrottle_ = 60000;
  /** @brief Time to wait between state checks */
  uint min_sleep_ = 500;

  // Derived private components
  /** @brief Files to get thermal data from */
  std::shared_ptr<io::FileCollection<u_long>> thermal_files_;
  /** @brief Files to get/set current CPU speed */
  std::shared_ptr<io::FileCollection<u_long>> scaling_max_freq_files_;
  /** @brief Files to get maximum CPU speed */
  std::shared_ptr<io::FileCollection<u_long>> cpuinfo_max_freq_files_;
  /** @brief Files to get mininum CPU speed */
  std::shared_ptr<io::FileCollection<u_long>> cpuinfo_min_freq_files_;
  /** @brief Files to get supported CPU throttle speeds */
  std::shared_ptr<io::FileCollection<std::string>>
      scaling_available_frequencies_files_;
  /** @brief File to get current CPU clock time information */
  std::shared_ptr<io::File<std::string>> proc_stat_file_;
  /** @brief Holds the vector of cpuinfo_max_freq values */
  std::vector<u_long> cpuinfo_max_freqs_;
  /** @brief Holds the vector of cpuinfo_min_freq values */
  std::vector<u_long> cpuinfo_min_freqs_;
  /** @brief Holds the matrix of available scaling frequencies */
  std::vector<std::vector<u_long>> scaling_available_frequencies_;

  // Internal functions

  // Asserts
  /**
   * @brief Asserts that SIGSTOP mode is enabled before accessing a given value
   *
   * @param value Value attempting to access
   * @throw templimiter::error::InternalError if SIGSTOP mode is not enabled
   */
  void assert_SIGSTOP_mode_(const std::string &value) const;

  /**
   * @brief Asserts that throttle mode is enabled before accessing a given value
   *
   * @param value Value attempting to access
   * @throw templimiter::error::InternalError if throttle mode is not enabled
   */
  void assert_throttle_mode_(const std::string &value) const;

  /**
   * @brief Asserts that at least one response mode is enabled
   *
   * @throw templimiter::error::ConfigError if neither mode is enabled
   */
  void assert_any_mode_() const;

  /**
   * @brief Asserts that throttle mode is gte dethrottle mode
   *
   * @throw templimiter::error::ConfigError if throttle temp is less than
   * dethrottle temp
   */
  void assert_throttle_gte_dethrottle_() const;

  /**
   * @brief Asserts that scaling_max_freq_files_ are sizey
   *
   * @param scalemax_sz Size of the scaling_max_freq_files_ object
   * @throw templimiter::error::ConfigError if files are not sizey
   */
  void assert_scale_max_freq_files_sizey_(size_t scalemax_sz) const;

  /**
   * @brief Asserts that generated scaling_available_frequencies_ matrix is
   * sizey
   *
   * @throw templimiter::error::ConfigError if files are not sizey
   */
  void assert_scaling_available_frequencies_sizey_() const;

  /**
   * @brief Asserts that scaling_max_freq_files_ size matches scaling available
   * matrix size
   *
   * @param scalemax_sz Size of the scaling_max_freq_files_ object
   * @throw templimiter::error::ConfigError if sizes are not equal
   */
  void assert_scaling_max_size_eq_scaling_available_(size_t scalemax_sz) const;

  /**
   * @brief Asserts that scaling_max_freq_files, cpuinfo_max_freq, and
   * cpuinfo_min_freq are the same size
   *
   * @param scalemax_sz Size of the scaling_max_freq_files_ object
   * @throw templimiter::error::ConfigError if cpu_max_freq_files_ size is zero
   * @throw templimiter::error::ConfigError if cpu_min_freq_files_ size is zero
   * @throw templimiter::error::ConfigError if scaling_max_freq_files_ size is
   * not equal to cpu_max_freq_files_ size
   * @throw templimiter::error::ConfigError if scaling_max_freq_files_ size is
   * not equal to cpu_min_freq_files_ size
   */
  void assert_scaling_max_size_eq_cpu_max_and_min_(size_t scalemax_sz) const;

  /**
   * @brief Asserts that SIGSTOP temp is gte SIGCONT temp
   *
   * @throws templimiter::error::ConfigError if SIGSTOP temp is less than
   * SIGCONT temp
   */
  void assert_SIGSTOP_gte_SIGCONT_() const;

  /**
   * @brief Asserts that proc_stat_file_ is sizey
   *
   * @param procstat_sz Size of the proc_stat_file_ object
   * @throws templimiter::error::InternalError if file is not sizey
   */
  void assert_proc_stat_file_sizey_(size_t procstat_sz) const;

  // Procedures
  /**
   * @brief Loads config lines to the private config_lines_
   * @param config_path Path to the config file
   */
  void load_config_lines_(const std::string &config_path);

  /** @brief Loads own pid from PROC_SELF_STAT_ to own_pid_ */
  void get_own_pid_();

  /**
   * @brief Retrieves the index of the tag in the config lines
   *
   * @param tag Tag to search for
   * @return size_t
   */
  std::vector<size_t> indices_of_tag_(const std::string &tag);

  /** @brief Loads all private config values from config lines */
  void load_config_values_();

  /** @brief Verifies all fallable conditions and loads file objects */
  void set_and_assert_config_();

  // Templates
  /**
   * @brief Loads a value from the config lines given a tag
   *
   * @tparam T Type of value
   * @param tag Tag to search for
   * @param defval Default value to return if not found
   * @return T
   * @throw templimiter::error::ConfigError if config has more than one line
   * configuring the tag
   * @throw templimiter::error::ConfigError if config line contains more than
   * one value for the tag
   * @throw templimiter::error::ConfigError if value type is invalid
   */
  template <typename T>
  T load_from_tag_(const std::string &tag, const T &defval) {
    // Try finding the index of the tag in the configuration
    std::vector<size_t> indices = indices_of_tag_(tag);
    if (indices.size() == 0) {
      // Tag is not found, return default
      io::log("Could not find tag <" + tag + "> in config. Using default");
      return defval;
    } else if (indices.size() > 1) {
      throw error::ConfigError(
          tag, "",
          "Multiple indices found for tag. Config should only have one value.");
    }

    // Set index to the only value in indices
    size_t index = indices[0];

    // Split the config line and take all values right of the tag
    std::vector<std::string> spl;
    std::vector<std::string> val_strs;
    spl = tools::split(config_lines_[index], ' ');
    val_strs = tools::subvect(spl, 1, ptrdiff_t(spl.size() - 1));

    // Ensure values have been found
    if (val_strs.size() < 1) {
      io::log("Could not find a value for tag <" + tag +
              "> in config. Using default");
      return defval;
    }

    // Ensure only one value is found
    if (val_strs.size() > 1) {
      throw error::ConfigError(
          tag, tools::join(val_strs, ", "),
          "Expected only one element in configuration setting.");
    }
    std::string val = val_strs.front();

    // Try to convert and return the value
    try {
      auto result = tools::convert<T>(val);
      return result;
    } catch (const error::Error &e) {
      if (e.name() == "TypeError") {
        throw error::ConfigError(tag, val, "Unable to convert type.");
      } else {
        throw;
      }
    }
  }

  /**
   * @brief Loads a vector value from the config lines given a tag
   *
   * @tparam T Type of the vector
   * @param tag Tag to search for
   * @param defval Default value to return if not found
   * @return T
   * @throw templimiter::error::ConfigError if config has more than one line
   * configuring the tag
   * @throw templimiter::error::ConfigError if value type is invalid
   */
  template <typename T>
  std::vector<T> load_from_tag_(const std::string &tag,
                                const std::vector<T> &defval) {
    // Try finding the index of the tag in the configuration
    std::vector<size_t> indices = indices_of_tag_(tag);
    if (indices.size() == 0) {
      // Tag is not found, return default
      io::log("Could not find tag <" + tag + "> in config. Using default");
      return defval;
    } else if (indices.size() > 1) {
      throw error::ConfigError(
          tag, "",
          "Multiple indices found for tag. Config should only have one value.");
    }

    // Set index to the only value in indices
    size_t index = indices[0];

    // Split the line and retrieve values right of tag
    std::vector<std::string> spl;
    std::vector<std::string> val_strs;
    spl = tools::split(config_lines_[index], ' ');
    val_strs = tools::subvect(spl, 1, ptrdiff_t(spl.size() - 1));
    if (val_strs.size() < 1) {
      io::log("Could not find any values for tag <" + tag +
              "> in config. Using default");
      return defval;
    }

    // Try to convert and return the value
    try {
      auto result = tools::convert<T>(val_strs);
      return result;
    } catch (const error::Error &e) {
      if (e.name() == "TypeError") {
        throw error::ConfigError(tag, tools::join(val_strs, ", "),
                                 "Unable to convert type.");
      } else {
        throw;
      }
    }
  }

 public:
  /**
   * @brief Construct a new Config object
   * @param config_path Where to find the configuration file
   */
  explicit Config(const std::string &config_path);

  /**
   * @brief Returns log_file_path configuration setting.
   * @return const std::string&
   */
  const std::string &log_file_path() const;

  /**
   * @brief Returns whitelist_pid configuration setting
   * @return const std::vector< pid_t >&
   */
  const std::vector<pid_t> &whitelist_pid() const;

  /**
   * @brief Returns whitelist_comm configuration setting
   * @return const std::vector< std::string >&
   */
  const std::vector<std::string> &whitelist_comm() const;

  /**
   * @brief Returns whitelist_state configuration setting
   * @return const std::vector< char >&
   */
  const std::vector<char> &whitelist_state() const;

  /**
   * @brief Returns whitelist_ppid configuration setting
   * @return const std::vector< pid_t >&
   */
  const std::vector<pid_t> &whitelist_ppid() const;

  /**
   * @brief Returns whitelist_pgrp configuration setting
   * @return const std::vector< int >&
   */
  const std::vector<int> &whitelist_pgrp() const;

  /**
   * @brief Returns whitelist_session configuration setting
   * @return const std::vector< int >&
   */
  const std::vector<int> &whitelist_session() const;

  /**
   * @brief Returns whitelist_tty_nr configuration setting
   * @return const std::vector< int >&
   */
  const std::vector<int> &whitelist_tty_nr() const;

  /**
   * @brief Returns whitelist_tpgid configuration setting
   * @return const std::vector< int >&
   */
  const std::vector<int> &whitelist_tpgid() const;

  /**
   * @brief Returns whitelist_flags configuration setting
   * @return const std::vector< uint >&
   */
  const std::vector<uint> &whitelist_flags() const;

  /**
   * @brief Returns whitelist_max_nice configuration setting
   * @return long
   */
  long whitelist_max_nice() const;

  /**
   * @brief Returns use_throttle configuration setting
   * @return true if using throttle
   * @return false if not using throttle
   */
  bool use_throttle() const;

  /**
   * @brief Returns use_SIGSTOP configuration setting
   * @return true if using SIGSTOP
   * @return false if not using SIGSTOP
   */
  bool use_SIGSTOP() const;

  /**
   * @brief Returns use_scaling_available configuration setting
   * @return true if using scaling_available
   * @return false if not using scaling_available
   */
  bool use_scaling_available() const;

  /**
   * @brief Returns the use_stepwise_SIGSTOP configuration setting
   *
   * @return true if using stepwise SIGSTOP
   * @return false if not using stepwise SIGSTOP
   */
  bool use_stepwise_SIGSTOP() const;

  /**
   * @brief Returns the use_stepwise_SIGCONT configuration setting
   *
   * @return true if using stepwise SIGCONT
   * @return false if not using stepwise SIGCONT
   */
  bool use_stepwise_SIGCONT() const;

  /**
   * @brief Returns temp_SIGSTOP configuration setting
   * @return u_long
   */
  u_long temp_SIGSTOP() const;

  /**
   * @brief Returns temp_SIGCONT configuration setting
   * @return u_long
   */
  u_long temp_SIGCONT() const;

  /**
   * @brief Returns temp_throttle configuration setting
   * @return u_long
   */
  u_long temp_throttle() const;

  /**
   * @brief Returns temp_dethrottle configuration setting
   * @return u_long
   */
  u_long temp_dethrottle() const;

  /**
   * @brief Returns min_sleep configuration setting
   * @return uint
   */
  uint min_sleep() const;

  /**
   * @brief Returns the constructed thermal_files FileCollection object based on
   * the configured matcher
   *
   * @return const std::shared_ptr< io::FileCollection< u_long > >&
   */
  const std::shared_ptr<io::FileCollection<u_long>> &thermal_files();

  /**
   * @brief Returns the constructed scaling_max_freq_files FileCollection object
   * based on the configured matcher
   *
   * @return const std::shared_ptr< io::FileCollection< u_long > >&
   */
  const std::shared_ptr<io::FileCollection<u_long>> &scaling_max_freq_files();

  /**
   * @brief Returns the constructed proc_stat_file File object based on the
   * hardcoded /proc/stat location
   *
   * @return const std::shared_ptr<io::File<std::string>>&
   */
  const std::shared_ptr<io::File<std::string>> &proc_stat_file();

  /**
   * @brief Returns the constructed vector of cpu max frequencies retrieved
   * during program initialization
   *
   * @return const std::vector< u_long >&
   */
  const std::vector<u_long> &cpuinfo_max_freqs() const;

  /**
   * @brief Returns the constructed vector of cpu min frequencies retrieved
   * during program initialization
   *
   * @return const std::vector< u_long >&
   */
  const std::vector<u_long> &cpuinfo_min_freqs() const;

  /**
   * @brief Returns the constructed vector of cpu available scaling frequencies
   * retrieved during program initialization
   *
   * @return const std::vector< u_long >&
   */
  const std::vector<std::vector<u_long>> &scaling_available_frequencies() const;
};

}  // namespace daemon

}  // namespace templimiter
